source/common/upstream/load_balancer_impl.h

```diff
@@ -70,10 +73,65 @@ class ZoneAwareLoadBalancerBase : public LoadBalancerBase {
                             const envoy::api::v2::Cluster::CommonLbConfig& common_config);
   ~ZoneAwareLoadBalancerBase();
 
+  // When deciding which hosts to use on an LB decision, we need to know how to index into the
+  // priority_set. This priority_set cursor is used by ZoneAwareLoadBalancerBase subclasses, e.g.
+  // RoundRobinLoadBalancer, to index into auxillary data structures specific to the LB for
+  // a given host set selection.
+  struct HostsSource {
+    enum class SourceType {
+      // All hosts in the host set.
+      AllHosts,
+      // All healthy hosts in the host set.
+      HealthyHosts,
+      // Healthy hosts for locality @ locality_index.
+      LocalityHealthyHosts,
+    };
+
+    HostsSource() {}
+
+    HostsSource(uint32_t priority, SourceType source_type)
+        : priority_(priority), source_type_(source_type) {
+      ASSERT(source_type == SourceType::AllHosts || source_type == SourceType::HealthyHosts);
+    }
+
+    HostsSource(uint32_t priority, SourceType source_type, uint32_t locality_index)
+        : priority_(priority), source_type_(source_type), locality_index_(locality_index) {
+      ASSERT(source_type == SourceType::LocalityHealthyHosts);
+    }
+
+    // Priority in PrioritySet.
+    uint32_t priority_{};
+
+    // How to index into HostSet for a given priority.
+    SourceType source_type_{};
+
+    // Locality index into HostsPerLocality for SourceType::LocalityHealthyHosts.
+    uint32_t locality_index_{};
+
+    bool operator==(const HostsSource& other) const {
+      return priority_ == other.priority_ && source_type_ == other.source_type_ &&
+             locality_index_ == other.locality_index_;
+    }
+  };
+
+  struct HostsSourceHash {
+    size_t operator()(const HostsSource& hs) const {
+      size_t hash = std::hash<uint32_t>()(hs.priority_);
```

Comment:
might be worth using xxhash here so this is stable across platforms.

Reply:
Don't think we need stability here, do we? We're not using this for anything other than indexing into a hash table, where it's just a performance hint to get buckets.

Reply:
Just if any tests relied on it. It's not a big deal either way but in general IMO we should probably always use a known hash function if possible.